RFCs/rfc-0004-fast-transactions.md

```diff
@@ -0,0 +1,415 @@
+
+Still, there is a technical possibility for zkApp to replace the transaction with the same nonce, not using the new state but using the old state, effectively rolling back the transactions not included yet in the Mina blockchain block.
+
+If we formulate what a guaranteed state means in the contract terms, it will lead us to the following wording:
+
+### **Terms of the State Guarantee Contract**
+
+- **Guarantee Issuer**: BlockSpace Node
+
+- **Beneficiary**: zkApp
+
+- **Condition Precedent**: zkApp is obliged at all times during the term of this agreement to maintain at its own expense the fee for the transactions that is necessary for avoiding the eviction of the transaction from the mempool and is enough for the inclusion of the transaction in the block during the maximum time that transaction is allowed to be in the mempool.
+
+- **Term**: The time that starts when zkApp sends the transaction to the BlockSpace Node to be committed to the L1 blockchain and ends on the earliest of:
+
+  - The transaction is included in the L1 block
+  - ZkApp calls off the transaction by sending the replacement transaction
+  - The transaction is excluded from the mempool due to the low fee or due to the maximum time reached for the transaction to be in the mempool as the result of the violations by zkApp of the Condition Precedent.
+
+- **Rights of Third Parties**: Third Parties, including, but not limited to, other zkApps and Accounts, are NOT beneficiaries of the State Guarantee Contract and can rely on this guaranteed state only in the optimistic sense by also relying on the architecture of the zkApp and not on this Guarantee. It is recommended for third parties to wait for BlockSpace to commit to the L1 before executing any material transaction that relies on the state, such as asset withdrawal.
+- **Applicable law**: Mina protocol
+- **Arbitrage**: O(1)Labs
+- **SLA**: TBD
+
+### The calculated state
+
+zkApp can calculate the state that will be the result of the transaction in a way similar to how Mina.transaction() does (or by using Mina.transaction() without calling transaction.prove())
+
+This calculated state can be used immediately by the zkApp to start creating and proving the next transaction without waiting for the previous transaction.prove() to be finished, enabling parallel proof calculations.
+
+The calculated state is not guaranteed and can be considered simply as one of the variables zkApp uses. To use it, Mina.transaction() arguments have to be amended to accept calculated or guaranteed state to be used as precondition input instead of the default precondition account state returned by fetchAccount() and referring to the last block account state.
+
+The calculated state will become a guaranteed state in an optimistic sense and, later, a final state in an optimistic sense.
```

Comment:
This, and the previous 2 sections, could be replaced with the sentence `We will optimistically assume that, whatever the BlockSpace node claims that the zkApp's state will be, is the new state that we should build upon.`

You can then use the space you've saved to describe how this will handle:
* any state changes while a zkApp proof is in flight, e.g.
  - txn1 requests the state, receives S1, begins proving
  - txn2 requests the state, receives S1, begins proving
  - txn1 finishes proving, submits, state becomes S2
  - txn2 finishes proving, its proof is now invalid.
  - (Scale this to the potential number of concurrent executions in the system.)
* cross-sequencing with other zkApps
  - Concretely, a zkApp may 'call' another zkApp, and thus depend on the linked state of the 2 zkApps.
  - This may become arbitrarily complex, if a zkApp is called by or calls may different other zkApps.
  - If 2 zkApps attempt to interact, but they are being sequenced by 2 different `BlockSpace` nodes, which node do they go to? How do those nodes coordinate? Do they coordinate? How does the block producer who might include one or other of the bundles know which to inlcude?
* communication
  - What do the queries look like from o1js to the `BlockSpace` node?
  - What is the incentive to serve those queries?
  - How does the `BlockSpace` node communicate with the upcoming block producer? (NB: that they are unknown until the block has been produced)
  - Concretely, running through multiple scenarios of racing transactions etc., run through which messages are sent to and from which entites at which time, what calculations they run, and what failure cases can occur at each of these interactions.

Reply:
@mrmr1993 
Thank you for your prompt answer and comments, which are always very valuable and to the point.

I have very limited documentation available for both Zeko and Protokit. Zeko is not available yet for testing, and Protokit is available with proofs off, so I prefer not to compare their features at this stage due to the lack of information.

This RFC has raised several issues for community discussion, and not all of those issues are technical. I believe that we should first look at the business objectives, and the technological decisions should be aligned with them.
- The thesis regarding the importance of fast, 1-second transactions for Mina protocol adoption. Is this of utmost importance for Mina protocol adoption, or merely a nice-to-have feature?
- If fast transactions are critical and required for Mina protocol adoption, and they are not built now in the protocol, can L2 solutions like Zeko and Protokit fully address this need? Or do we need to add it to the protocol or have a broader range of solutions aiming to have at least three solutions for fast transactions?
- Zeko is a full-fledged L2 with bridge, DA layer, explorer, etc. Do we need a “light” ephemeral blockspace that will complement this L2 solution or not?

Depending on the consensus on answers to the above questions, we must continue to find technical design decisions for light ephemeral blockspace or other solutions for fast transactions and integrate it into the Mina protocol or make it a separate solution, with this discussion on the RFC being a good start, or we may close this topic if it's deemed non-essential for Mina protocol adoption.

I would highly appreciate your opinion on the above questions as you undoubtedly have more information and insights on those topics, and your opinion is very valuable for the community.

Reply:
@dfstio Protokit isn't only available with 'proofs off', we simply did not put any emphasis on the feature yet since we haven't tested the entire framework with proofs enabled end to end. But to address this i wrote a test that runs a sample Protokit runtime `Balances` in the prover. It takes roughly 20s to prove a circuit of < 500 gates, this circuit is an implementation for `transfer()` which internally handles 4 state transitions, uses `this.transaction.sender` and `assert()` APIs of Protokit (end to end provable). Furthermore the state transitions need to be proven in parallel, and will be proven in batches of 4 in state transition circuits that almost exhaust the 2^16 constraint limit. Proving a 2^16 circuit in wasm takes about 50s on my machine. 

The way Protokit is architectured allows running both provers in paralel, therefore it takes about 50s combined to prove  + then to merge the proofs of runtime execution and state transition application (2x verify() in our BlockProver).

You don't have to write any other code than your runtime business logic, rest of what i've described including sequencing, mempool, proof paralelization / task input precomputation etc is handled internally by the framework.

Here's the code for the Balances runtime:

```typescript

@runtimeModule()
class Balances extends RuntimeModule<Record<string, never>> {
  @state() public balances = StateMap.from<PublicKey, UInt64>(
    PublicKey,
    UInt64
  );

  @runtimeMethod()
  public transfer(from: PublicKey, to: PublicKey, amount: UInt64) {
    // read current balances
    const fromBalance = this.balances.get(from).value;
    const toBalance = this.balances.get(to).value;

    const fromIsSender = this.transaction.sender.equals(from);
    // check if from is the sender
    assert(fromIsSender, "Sender mismatch");

    // check if balance is sufficient
    const fromBalanceIsSufficient = fromBalance.greaterThanOrEqual(amount);
    assert(fromBalanceIsSufficient, "Insufficient balance");

    // naive implementation of 'safe sub' thats circuit friendly
    const paddedFrombalance = fromBalance.add(amount);
    const safeFromBalance = Provable.if<UInt64>(
      fromBalanceIsSufficient,
      UInt64,
      fromBalance,
      paddedFrombalance
    );

    // calculate new balances
    const newFromBalance = safeFromBalance.sub(amount);
    const newToBalance = toBalance.add(amount);

    this.balances.set(from, newFromBalance);
    this.balances.set(to, newToBalance);
  }
}
```

You can find the complete test for the runtime in-prover in our repository:
https://github.com/proto-kit/framework/blob/feature/runtime-zk-programmable/packages/module/test/RuntimeZkProgrammable.test.ts#L220

P.S.: We're actively working on documentation, in the meantime the entire repository is open source and can be learned from e.g. by checking out the provided tests and examples. I think its not exactly productive for everybody involved to write a whole RFC describing a solution that's by design worse off than the existing ones, with the main argument being "docs dont exists" (for both). If there's anything missing i'd be more than happy to explain in any format desirable (office hours, calls, workshops) until the full docs are ready.